core/src/test/scala/unit/kafka/log/LogValidatorTest.scala

```diff
@@ -773,6 +773,23 @@ class LogValidatorTest {
       partitionLeaderEpoch = RecordBatch.NO_PARTITION_LEADER_EPOCH)
   }
 
+  @Test(expected = classOf[InvalidRecordKeyException])
+  def testNullKeyForCompactedTopic() {
```

Comment:
`// throw a non retriable exception only if the produce request is V3 or higher`

i.e., we could add a negative test that ensures we haven't, and won't in the future, change the semantics of this for requests <V3